client/packages/programs/src/JsonForms/components/Search/SearchWithUserSource.tsx

```diff
@@ -0,0 +1,173 @@
+  const { data, path, handleChange, label, uischema } = props;
+  const { errors: zErrors, options } = useZodOptionsValidation(
+    Options,
+    uischema.options
+  );
+  const t = useTranslation('programs');
+  const [searchText, setSearchText] = useState('');
+  const [editMode, setEditMode] = useState(!data);
+  const [noResultsText, setNoResultsText] = useState(
+    t('control.search.searching-label')
+  );
+
+  const {
+    runQuery,
+    getOptionLabel,
+    getDisplayElement,
+    saveFields,
+    placeholderText,
+    loading,
+    error: queryError,
+    results,
+  } = useSearchQueries(options ?? {});
+
+  const debouncedOnChange = useDebounceCallback(
+    value => {
+      if (value.length >= MIN_CHARS) runQuery(value);
+      else {
+        // Clear the results if user input falls *below* `minChars`
+        if (results.length) runQuery('');
+      }
+      setNoResultsText(t('control.search.no-results-label'));
+    },
+    [searchText],
+    500
+  );
+
+  const handleDataUpdate = (selectedResult: Record<string, any> | null) => {
+    if (selectedResult === null) return;
+    if (!saveFields) handleChange(path, selectedResult);
+
+    const newObj: Record<string, any> = {};
+    saveFields?.forEach(
+      field => (newObj[field] = selectedResult[field] ?? null)
+    );
+    handleChange(path, newObj);
+    setEditMode(false);
+  };
+
+  if (!options) return null;
+
+  return (
+    <Box
+      display="flex"
+      alignItems="center"
+      gap={2}
+      style={{ minWidth: 300 }}
+      margin={0.5}
+      marginLeft={0}
+    >
+      <Box style={{ textAlign: 'end' }} flexBasis={FORM_LABEL_COLUMN_WIDTH}>
+        <FormLabel sx={{ fontWeight: 'bold' }}>{label}:</FormLabel>
+      </Box>
+      {editMode ? (
+        <Box flexBasis={FORM_INPUT_COLUMN_WIDTH} justifyContent="flex-start">
+          <>
+            <Autocomplete
+              sx={{ '.MuiFormControl-root': { minWidth: '100%' } }}
+              options={results}
+              disabled={!props.enabled}
+              onChange={(_, option) => handleDataUpdate(option)}
+              onInputChange={(_, value) => {
+                debouncedOnChange(value);
+                setSearchText(value);
+                setNoResultsText(t('control.search.searching-label'));
+              }}
+              onBlur={() => {
+                if (data) setEditMode(false);
+              }}
+              getOptionLabel={getOptionLabel ?? undefined}
+              clearable={!props.config?.required}
+              inputProps={{
+                error: !!zErrors || !!props.errors || !!queryError,
+                helperText: zErrors ?? props.errors ?? queryError,
+              }}
+              noOptionsText={
+                loading
+                  ? t('control.search.searching-label')
+                  : searchText.length < MIN_CHARS
+                  ? t('control.search.below-min-chars', { minChars: MIN_CHARS })
+                  : noResultsText
+              }
```

Comment:
even when laid out nicely like this, I think that nested ternary operators are evil. It's too hard to understand what the intention is.

```suggestion

  const getNoOptionsText = () => {
    switch (true) {
      case loading:
        return t('control.search.searching-label');
      case searchText.length < MIN_CHARS:
        return t('control.search.below-min-chars', { minChars: MIN_CHARS });
      default:
        return noResultsText;
    }
  };
  ...
              noOptionsText={getNoOptionsText()}
```